Source/Core/Common/EnumFormatter.h

```diff
@@ -0,0 +1,86 @@
+// Copyright 2021 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <array>
+#include <fmt/format.h>
+#include <type_traits>
+
+/*
+ * Helper for using enums with fmt.
+ *
+ * Usage example:
+ *
+ * enum class Foo
+ * {
+ *   A = 0,
+ *   B = 1,
+ *   C = 2,
+ * };
+ *
+ * template <>
+ * struct fmt::formatter<Foo> : EnumFormatter<Foo::C>
+ * {
+ *   formatter() : EnumFormatter({"A", "B", "C"}) {}
+ * };
+ *
+ * enum class Bar
+ * {
+ *   D = 0,
+ *   E = 1,
+ *   F = 3,
+ * };
+ *
+ * template <>
+ * struct fmt::formatter<Bar> : EnumFormatter<Bar::F>
+ * {
+ *   // using std::array here fails due to nullptr not being const char*, at least in MSVC
+ *   // (but only when a field is used; directly in the constructor is OK)
+ *   static constexpr array_type names = {"D", "E", nullptr, "F"};
+ *   formatter() : EnumFormatter(names) {}
+ * };
+ */
+template <auto last_member, typename T = decltype(last_member),
+          size_t size = static_cast<size_t>(last_member) + 1,
+          std::enable_if_t<std::is_enum_v<T>, bool> = true>
+struct EnumFormatter
+{
+  // This is needed because std::array deduces incorrectly if nullptr is included in the list
+  using array_type = std::array<const char*, size>;
+  const array_type m_names;
+  bool formatting_for_shader = false;
+  EnumFormatter(const array_type names) : m_names(names) {}
```

Comment:
explicit?

Also, if this wasn't const, you could use `std::move` on the names

Reply:
`explicit` is probably a good idea, and the constructor should probably also be `protected` (and the fields also `private`) to more clearly convey how it should be used.  I think the constructor could also be `constexpr` but it doesn't look liek that really makes a difference.

`std::array` seems to support move construction even when const (at least when it's an array of `const char *`), but looking at compiler explorer, it doesn't seem to make a difference here.  I do think undue copies are a mild concern (I'm not 100% sure whether fmt constructs it each time and thus a new array is created each time, or if the array is constructed once, or what exactly); theoretically, I could instead have it take a const reference to a static constexpr field, but that would mean that a field is always needed.